ir/function.cpp

```diff
@@ -293,12 +362,129 @@ bool Function::removeUnusedStuff(const multimap<Value*, Value*> &users,
+    worklist.emplace_back(root, false);
+  }
+
+  while (!worklist.empty()) {
+    auto &[bb, flag]  = worklist.back();
+    if (flag) {
+      worklist.pop_back();
+      map<Value*, Value*> vmap;
+      map<const BasicBlock *, vector<const BasicBlock *>> bbmap;
+      vector<const BasicBlock *> cloned;
+      auto bb_nodes = Nodes[bb];
+
+      for (auto &&bb_ii : bb_nodes) {
+        bbmap[bb_ii].push_back(bb_ii);
+      }
+      for (unsigned unroll = 1 ; unroll <= k; ++unroll) {
+        for (auto &&bb_ii : bb_nodes) {
+          auto &bb_i = const_cast<BasicBlock&>(*bb_ii);
+          string twine = "_" + string(bb_dot_name(bb->getName())) +
+                         "_" + to_string(unroll);
+          auto bb_cloned = &cloneBB(bb_i, twine, vmap);
+          bbmap[bb_ii].push_back(bb_cloned);
+          cloned.push_back(bb_cloned);
+        }
+      }
+
+      vector<const BasicBlock *> backedge_srcs;
+      set<const BasicBlock *> nodes_set(bb_nodes.begin(), bb_nodes.end());
+      // rewire targets
+      for (unsigned unroll = 0 ; unroll <= k; ++unroll) {
+        for (auto &&bb_ii : bb_nodes) {
+          auto curr = const_cast<BasicBlock *>(bbmap[bb_ii][unroll]);
+          map<const BasicBlock *, const BasicBlock *> replacement;
+          for (auto &tgt : curr->targets()) {
+            if (&tgt == bb || bb_ii != bb)
+              backedge_srcs.push_back(bb_ii);
+            // Filter out exit edges.
+            if (!nodes_set.count(&tgt))
+              continue;
+
+            const BasicBlock *to = nullptr;
+            // backedge
+            if (&tgt == bb) {
+              if (unroll == k)
+                to = &sink;
+              else
+                to = bbmap[&tgt][unroll + 1];
+            }
+            // edges inside loop
+            else {
+              to = bbmap[&tgt][unroll];
+            }
+            replacement[&tgt] = to;
+          }
+          for (auto &[from, to] : replacement)
+            curr->replaceTargetWith(from, to);
+        }
+      }
+
+      // rewire sources on header
+      auto bb_c = const_cast<BasicBlock*>(bb);
+
+      for (auto &src : backedge_srcs) {
+        bb_c->prunePhiSrc(src);
+      }
+
+      // rewire sources on nonheaders
+      for (unsigned unroll = 1; unroll <= k; ++unroll) {
+        auto work_bb = const_cast<BasicBlock *>(bbmap[bb][unroll]);
+        work_bb->prunePhiSrcsExcept(backedge_srcs);
+        for (auto &src : backedge_srcs)
+          work_bb->replacePhiSrcWith(src, bbmap[src][unroll - 1]);
+
+        for (auto &&bb_ii : bb_nodes) {
+          if (bb_ii == bb)
+            continue;
+          auto curr = const_cast<BasicBlock *>(bbmap[bb_ii][unroll]);
+          for (auto &src : bb_nodes)
+            curr->replacePhiSrcWith(src, bbmap[src][unroll]);
+        }
+      }
+
+      // add clones back to all the loop bb's ancestors
+      const BasicBlock *root = bb;
+      while (bottomup.count(root)) {
+        auto next = bottomup.at(root);
```

Comment:
Instead of using direct operations on `bottomup`, what about adding `LoopAnalysis::getParent()` and using it?

Reply:
Or, maybe we can move this while loop into a new `LoopAnalysis::updateTree` (not a strong opinion though)?